### PR TITLE
Register KEB as an integration system

### PR DESCRIPTION
**Pull request by polskikiel**
**Description**

Changes proposed in this pull request:

- Register KEB as an integration system

**Related issue(s)**

### CODE REVIEW

File: chart/compass/charts/kyma-environment-broker/templates/deployment.yaml [hunk 1]
```diff
@@ -82,6 +90,14 @@ spec:
                 secretKeyRef:
                   name: compass-postgresql
                   key: postgresql-sslMode
+            - name: APP_DEFAULT_TENANT
+              value: "{{ .Values.global.defaultTenant }}"
+            - name: APP_DIRECTOR_URL
+              value: "https://{{ .Values.global.gateway.tls.secure.oauth.host }}.{{ .Values.global.ingress.domainName }}/director/graphql"
```
Comment (jasiu001) [inline chart/compass/charts/kyma-environment-broker/templates/deployment.yaml]: `https://compass-gateway-auth-oauth.kyma.local/director/graphql` is it correct URL?
what about port? where are `namespace` and `svc` parts?

File: chart/compass/templates/registration-job.yaml [hunk 2]
```diff
@@ -60,7 +60,8 @@ spec:
               TOKEN_PAYLOAD='{"scopes": "'${SCOPES}'","tenant":"{{ .Values.global.defaultTenant }}"}'
```
Comment (jasiu001) [inline chart/compass/templates/registration-job.yaml]: Why is it moved from `provisioner` chart? why bash?

File: components/kyma-environment-broker/Gopkg.toml [hunk 3]
```diff
@@ -85,6 +85,10 @@
   name = "github.com/vrischmann/envconfig"
   version = "1.2.0"
 
+[[constraint]]
+  name = "github.com/kubernetes/client-go"
+  version = "kubernetes-1.15.6"
```
Comment (jasiu001) [inline components/kyma-environment-broker/Gopkg.toml]: why so old? last is `1.17.1`

File: chart/compass/charts/kyma-environment-broker/templates/deployment.yaml [hunk 4]
```diff
@@ -45,13 +55,23 @@ spec:
...
               valueFrom:
                 secretKeyRef:
                   name: {{ include "kyma-env-broker.fullname" . }}
                   key: broker-password
+            - name: APP_PROCESS_RUNTIME_ID
+              value: "{{ .Values.provisioner.processRuntimeID }}"
+            - name: APP_DIRECTOR_NAMESPACE
+              value: "{{ .Release.Namespace }}"
+            - name: APP_DIRECTOR_DEFAULT_TENANT
+              value: "{{ .Values.global.defaultTenant }}"
```
Comment (jasiu001) [inline chart/compass/charts/kyma-environment-broker/templates/deployment.yaml]: Do we need `APP_DIRECTOR_DEFAULT_TENANT` as a parameter?

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 5]
```diff
@@ -55,18 +72,57 @@ func main() {
 		provisionerClient = provisioner.NewProvisionerClient(cfg.Provisioning.URL, true)
 	}
 
+	secrets, err := newSecretsInterface(cfg.Director.Namespace)
+	fatalOnError(err)
+	oauthClient := oauth.NewOauthClient(newHTTPClient(false), secrets, cfg.Director.OauthCredentialsSecretName)
+	fatalOnError(oauthClient.WaitForCredentials())
+	tkn, err := oauthClient.GetAuthorizationToken()
+	fatalOnError(err)
+	fmt.Println("DUPA:", tkn.AccessToken)
```
Comment (jasiu001) [inline components/kyma-environment-broker/cmd/broker/main.go]: :)

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 6]
```diff
@@ -55,18 +72,57 @@ func main() {
...
+		if err != nil {
+			return nil, errors.Errorf("failed to read k8s in-cluster configuration, %s", err.Error())
+		}
+	}
+
+	coreClientset, err := kubernetes.NewForConfig(k8sConfig)
+	if err != nil {
+		return nil, errors.Errorf("failed to create k8s core client, %s", err.Error())
+	}
+
```
Comment (jasiu001) [inline components/kyma-environment-broker/cmd/broker/main.go]: Is it not easier to use:
```
cfg, err := config.GetConfig()
k8sCli, err := client.New(cfg, client.Options{})
```
from:
"sigs.k8s.io/controller-runtime/pkg/client"
"sigs.k8s.io/controller-runtime/pkg/client/config"
libraries?

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 7]
```diff
@@ -55,18 +72,57 @@ func main() {
...
+func newHTTPClient(skipCertVeryfication bool) *http.Client {
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: skipCertVeryfication},
+		},
+		Timeout: 30 * time.Second,
+	}
+}
+
+func newSecretsInterface(namespace string) (v1.SecretInterface, error) {
```
Comment (jasiu001) [inline components/kyma-environment-broker/cmd/broker/main.go]: I am not convinced about `interface` part in the function name :)

File: chart/compass/charts/kyma-environment-broker/templates/deployment.yaml (hunk 1 above)
Comment (polskikiel) [reply chart/compass/charts/kyma-environment-broker/templates/deployment.yaml]: changed to `http://compass-director.compass-system.svc.cluster.local:3000/graphql`

File: chart/compass/templates/registration-job.yaml (hunk 2 above)
Comment (polskikiel) [reply chart/compass/templates/registration-job.yaml]: We have agreed with PK to move it to the chart's root, use its codebase and only create a separate secret for KEB. I would prefer to have it written in Go.

File: chart/compass/charts/kyma-environment-broker/templates/deployment.yaml (hunk 4 above)
Comment (polskikiel) [reply chart/compass/charts/kyma-environment-broker/templates/deployment.yaml]: we don't, good catch

File: components/kyma-environment-broker/cmd/broker/main.go (hunk 5 above)
Comment (polskikiel) [reply components/kyma-environment-broker/cmd/broker/main.go]: easy way to test this ;p

File: components/kyma-environment-broker/cmd/broker/main.go (hunk 7 above)
Comment (polskikiel) [reply components/kyma-environment-broker/cmd/broker/main.go]: ok will change to client

File: chart/compass/templates/registration-job.yaml [hunk 8]
```diff
@@ -141,12 +142,71 @@ spec:
...
+                  -H 'Content-Type: application/json' \
+                  -H 'Accept: application/json' \
+                  -H 'Connection: keep-alive' \
+                  -H "authorization: Bearer ${INTERNAL_TOKEN}" \
+                  --data-binary '{"query":"mutation requestCredentials {\n\trequestClientCredentialsForIntegrationSystem (id : \"'"$INTEGRATION_SYSTEM_ID"'\") {\n    id\n    auth {\n      credential {\n        ... on BasicCredentialData {\n            username\n            password\n        }\n        ...  on OAuthCredentialData {\n            clientId\n            clientSecret\n            url\n        }\n      }\n    }\n  }\n}\n"
+                 }')
+
+                CLIENT_ID=$(echo $CREDENTIAL_RESPONSE_BODY | jq -e '.data .requestClientCredentialsForIntegrationSystem .auth .credential .clientId')
+                CLIENT_SECRET=$(echo $CREDENTIAL_RESPONSE_BODY | jq -e '.data .requestClientCredentialsForIntegrationSystem .auth .credential .clientSecret')
+                TOKENS_ENDPOINT=$(echo $CREDENTIAL_RESPONSE_BODY | jq -e '.data .requestClientCredentialsForIntegrationSystem .auth .credential .url')
```
Comment (Szymongib) [inline chart/compass/templates/registration-job.yaml]: We duplicate a lot of code here. Maybe we could create a function `registerIntegrationSystem(name, description, secretName)` and just use it twice, once for Provisioner and once for Broker?

File: chart/compass/values.yaml [hunk 9]
```diff
@@ -79,9 +79,20 @@ global:
...
+
   provisioner:
     graphql:
       port: 3000
+    secrets:
+      integrationSystemCredentials:
+        name: compass-provisioner-credentials
+
+  oauth2:
+    host: oauth2.kyma.local
```
Comment (Szymongib) [inline chart/compass/values.yaml]: Maybe we can wrap it under `local` like this:
```
global
  localEnv:
    oauth2:
        host: ...
```
To signify it is used only on local environment. Or preferably put only a first part of host like:
```
global
  oauth2:
      host: oauth2
```
and use `.{{ .Values.global.ingress.domainName }}` to complete the full domain.

File: chart/compass/templates/registration-job.yaml (hunk 8 above)
Comment (polskikiel) [reply chart/compass/templates/registration-job.yaml]: You are right, I created shared methods:
`ensure_secret(name, description, secretName)`
`register_integration_system(name, description, secretName)`

and they are executed like that:
```
              echo "Checking if provisioner secret with credentials exists..."
              ensure_secret "compass-provisioner" "Compass Provisioner" $PROVISIONER_SECRET_NAME

              echo "Checking if broker secret with credentials exists..."
              ensure_secret "compass-kyma-environment-broker" "Compass Kyma Environment Broker" $BROKER_SECRET_NAME
```

**Pull request merged**